lib/urlapi.c

```diff
@@ -1307,6 +1307,178 @@ CURLU *curl_url_dup(const CURLU *in)
+                      ((x) == '+'))
+
+static CURLUcode normalize_query(const char *str, char **updated)
+{
+  /* to handle ' ' to '+' escaping we cannot use libcurl's URL encode
+     function */
+  struct dynbuf dupe;
+  size_t qlen = strlen(str);
+  CURLcode result = CURLE_OK;
+  if(!qlen) {
+    *updated = strdup("");
+    return *updated ? CURLUE_OK : CURLUE_OUT_OF_MEMORY;
+  }
+  Curl_dyn_init(&dupe, qlen * 3);
+
+  while(!result && qlen--) {
+    /* decode then encode */
+
+    /* treat the characters unsigned */
+    unsigned char in = (unsigned char)*str++;
+
+    if(('%' == in) && (qlen >= 2) && ISXDIGIT(str[0]) && ISXDIGIT(str[1])) {
+      curl_off_t byte;
+      char buf[4] = {'%'};
+      const char *p = &buf[1];
+      buf[1] = Curl_raw_toupper(str[0]);
+      buf[2] = Curl_raw_toupper(str[1]);
+      buf[3] = 0;
+      (void)Curl_str_hex(&p, &byte, 255);
+      if(!ISQUERYOK(byte) && (byte > 0x20) && (byte < 0x7f)) {
+        unsigned char num = (unsigned char)byte;
+        result = Curl_dyn_addn(&dupe, &num, 1);
+      }
+      else
+        /* add the original hex in uppercase */
+        result = Curl_dyn_addn(&dupe, buf, 3);
+      str += 2;
+      qlen -= 2;
+      continue;
+    }
+
+    if(in == ' ')
+      result = Curl_dyn_addn(&dupe, "+", 1);
+    else if(ISALNUM(in) || ISURLPUNTCS(in) || ISQUERYOK(in))
+      result = Curl_dyn_addn(&dupe, &in, 1);
+    else {
+      /* encode it */
+      const char hex[] = "0123456789abcdef";
+      char out[3];
+      out[0]='%';
+      out[1] = hex[in >> 4];
+      out[2] = hex[in & 0xf];
+      result = Curl_dyn_addn(&dupe, out, 3);
+    }
+  }
+  if(result)
+    return CURLUE_OUT_OF_MEMORY;
+  *updated = Curl_dyn_ptr(&dupe);
+  return CURLUE_OK;
+}
+
+/* NOTE: does not work for the query */
+static CURLUcode normalize_part(const char *ptr, char **updated)
+{
+  if(ptr) {
+    int olen;
+    char *uptr;
+    size_t ptrlen = strlen(ptr);
+
+    /* First URL decode the component */
+    char *rawptr = curl_easy_unescape(NULL, ptr, (int)ptrlen, &olen);
+    if(!rawptr)
+      return CURLUE_OUT_OF_MEMORY;
+
+    /* Then URL encode it again */
+    uptr = curl_easy_escape(NULL, rawptr, olen);
+    free(rawptr);
+    if(!uptr)
+      return CURLUE_OUT_OF_MEMORY;
+
+    *updated = uptr;
+  }
+  return CURLUE_OK;
+}
+
+#define MAX_HOSTNAME_LENGTH 1025
+
+/* if there is uppercase ASCII in the hostname, go lowercase */
+static CURLUcode normalize_host(const char *name, char **oname)
+{
+  struct dynbuf host;
+  char c;
+  CURLcode result = CURLE_OK;
+  int i = 0;
+  Curl_dyn_init(&host, MAX_HOSTNAME_LENGTH);
+  while(!result && (c = name[i++])) {
+    if(ISUPPER(c)) {
+      char l = Curl_raw_tolower(c);
+      result = Curl_dyn_addn(&host, &l, 1);
+    }
+    else
+      result = Curl_dyn_addn(&host, &c, 1);
+  }
+  if(result)
+    return CURLUE_OUT_OF_MEMORY;
+
+  /* a zero length string might return NULL here */
+  *oname = Curl_dyn_ptr(&host);
+  return CURLUE_OK;
+}
```

Comment:
Is this guaranteed to already have only characters valid in a host (sub-delims, unreserved, and %-encoded)?  Is punycode encoding and decoding guaranteed to result in something valid for this?  Having an unescaped `%`, `@`, or `/` in the output would be bad.

Reply:
Yes, the detection and rejection of illegal characters is done already when a URL or a host component is set.